kmod/src/data.c

```diff
@@ -1565,48 +1563,95 @@ int scoutfs_data_fiemap(struct inode *inode, struct fiemap_extent_info *fieinfo,
 	iblock = start >> SCOUTFS_BLOCK_SM_SHIFT;
 	last = (start + len - 1) >> SCOUTFS_BLOCK_SM_SHIFT;
 
+	args.ino = ino;
+	args.inode = inode;
+
+	/* outer loop */
 	while (iblock <= last) {
-		ret = scoutfs_ext_next(sb, &data_ext_ops, &args,
-				       iblock, 1, &ext);
-		if (ret < 0) {
-			if (ret == -ENOENT)
-				ret = 0;
-			last_flags = FIEMAP_EXTENT_LAST;
+		/* lock */
+		inode_lock(inode);
+		down_read(&si->extent_sem);
+
+		ret = scoutfs_lock_inode(sb, SCOUTFS_LOCK_READ, 0, inode, &lock);
+		if (ret) {
+			lock = NULL;
 			break;
 		}
 
-		trace_scoutfs_data_fiemap_extent(sb, ino, &ext);
+		args.lock = lock;
 
-		if (ext.start > last) {
-			/* not setting _LAST, it's for end of file */
-			ret = 0;
-			break;
+		/* collect entries */
+		info = page_address(page);
+		while (entries < (PAGE_SIZE / sizeof(struct fiemap_extent)) - 1) {
+			ret = scoutfs_ext_next(sb, &data_ext_ops, &args,
+					       iblock, 1, &ext);
+			if (ret < 0) {
+				if (ret == -ENOENT)
+					ret = 0;
+				last_flags = FIEMAP_EXTENT_LAST;
+				break;
+			}
+
+			trace_scoutfs_data_fiemap_extent(sb, ino, &ext);
+
+			if (ext.start > last) {
+				/* not setting _LAST, it's for end of file */
```

Comment:
This can now loop forever 'cause it's not breaking out of the outer iblock < last.  You'd need to set iblock here to break out (or have a complete like the other loops).  I guess the same's true of the ENOENT->LAST break as well (once we fix the later last_flags test and break).

Reply:
I guess this is going to be easier to get consistent with a complete flag. At least to me.